optaplanner-examples/src/main/resources/org/optaplanner/examples/rocktour/solver/rockTourScoreRules.drl

```diff
@@ -22,72 +22,66 @@ import org.optaplanner.core.api.score.buildin.hardmediumsoftlong.HardMediumSoftL
     then
-        scoreHolder.addHardConstraintMatch(kcontext, -1000);
+        scoreHolder.penalize(kcontext);
 end
 
-rule "Schedule as many shows as possible"
+rule "Unassigned show"
     when
-        RockTourParametrization($weight : missedShowPenalty != 0)
         RockShow(bus != null, date == null)
     then
-        scoreHolder.addHardConstraintMatch(kcontext, - $weight);
+        scoreHolder.penalize(kcontext);
 end
 
 // ############################################################################
-// Medium constraints
+// Medium constraints usually
 // ############################################################################
 
-rule "Maximize revenue opportunity"
+rule "Revenue opportunity"
     when
-        RockTourParametrization($weight : revenueOpportunity != 0)
         RockShow(date != null, $revenueOpportunity : revenueOpportunity)
     then
-        scoreHolder.addMediumConstraintMatch(kcontext, $weight * $revenueOpportunity);
+        scoreHolder.reward(kcontext, $revenueOpportunity);
 end
 
-rule "Minimize driving time"
+rule "Driving time to show per second"
     when
-        RockTourParametrization($weight : drivingTimeCostPerSecond != 0)
         RockShow(previousStandstill != null, $drivingTime : drivingTimeFromPreviousStandstill)
     then
-        scoreHolder.addMediumConstraintMatch(kcontext, - $weight * $drivingTime);
+        scoreHolder.penalize(kcontext, $drivingTime);
 end
 
-//rule "Minimize driving time to bus ending location"
-//    when
-//        RockTourParametrization($weight : drivingTimeCostPerSecond != 0)
-//        RockShow(nextShow == null, bus != null, $drivingTime : drivingTimeToBusArrivalLocation)
-//    then
-//        scoreHolder.addSoftConstraintMatch(kcontext, - $weight * $drivingTime);
-//end
+rule "Driving time to bus arrival per second"
+    when
+        RockShow(nextShow == null, bus != null, $drivingTime : drivingTimeToBusArrivalLocation)
+    then
+        scoreHolder.penalize(kcontext, $drivingTime);
+end
 
-rule "Visit sooner than later"
+rule "Delay show cost per day"
     when
-        RockTourParametrization($weight : delayCostPerDay != 0)
         RockShow(date != null, $daysAfterBusDeparture : daysAfterBusDeparture)
     then
-        scoreHolder.addMediumConstraintMatch(kcontext, - $weight * $daysAfterBusDeparture);
+        scoreHolder.penalize(kcontext, $daysAfterBusDeparture);
 end
 
 // ############################################################################
-// Soft constraints
+// Soft constraints usually
 // ############################################################################
 
-rule "Avoid long driving times"
+rule "Shorten driving time per millisecond squared"
     when
         RockShow(previousStandstill != null, $drivingTime : drivingTimeFromPreviousStandstill)
     then
-        scoreHolder.addSoftConstraintMatch(kcontext, - $drivingTime* $drivingTime);
+        scoreHolder.penalize(kcontext, $drivingTime * $drivingTime);
```

Comment:
@ge0ffrey This is a bit outside the regular PR review but why don't we generalize the concept of constraint weight to a constraint weight *function*, so for example:
```java
// simpler, allows a second parameter (we can have BiFunction)
scoreHolder.penalize(kcontext, $drivingTime, [optional 2nd param]);

@ConstraintWeight(SHORTEN_DRIVING_TIME_PER_MILLISECOND_SQUARED)
private Function<Long, HardMediumSoftLongScore> shortenshortenDrivingTimePerMillisecondSquared
        // any kind of "magic"
        = drivingTime -> HardMediumSoftLongScore.ofSoft(drivingTime * drivingTime);
```

Just an idea...

Reply:
I like this idea, but introducing it makes the boundary between constraint rules and weights blurry. You can do the same in rules and even combine it (maybe unintentionally).

Reply:
I wouldn't be worried about duplicate calculations - users can and should use ScoreVerifier to unit test their rules.

The main problem with this idea, I think, is that it might become difficult to change the weight from a UI.

Reply:
Interesting idea.
I definitely want to support "simple" @ConstraintWeights that are just a score too. Even if it's only because it's a PITA to store lamdba's in a database. Notice that shortenshortenDrivingTimePerMillisecondSquared is probably stored in a column in a table in the database.

I also did a performance trick to "unravel" the score weight multiplication if there is only one non-zero score level. That trick wouldn't be possible here, but that unravel doesn't happen if the score weight is for example 0hard/1medium/1soft (2 non-zero's) either (but this is a rare weight).

So, this is an interesting RFE, good for a separate jira.

Ps: note that ConstraintStreams will also support a lamdba parameter in the penalize() and reward() methods, but they 'll be used to extract the weightMultiplier from the matched facts. So that's unrelated.

Reply:
Yes, weight persistence is an important requirement to keep in mind. Didn't realize that.